### ISSUE TITLE
Understand Training Times for a 100M parameter model

### ISSUE THREAD

**Issue by Plyb** [DRAGNLabs/MIRROR-Pipeline#266]
**Acceptance Criteria**:
- we have a good understanding of what the best configuration for training a 100M parameter model on the supercomputer is (taking into account that some devices will have more demand than others)
  - How many nodes/gpus
  - what kinds of gpus
  - What Fabric strategy

Note: Dr. Fulda mentioned there may be things we can do to get higher supercomputer priority.

#252 has some thoughts on how to train efficiently at scale

Some things to maybe try:
- How does strategy change things? skipped - good enough
- What is the best kind of gpu we can use? skipped - good enough
- training example to token ratio? 2M tokens/100k examples for church translation (~20tok/ex)
- torch.set_float32_matmul_precision('medium' | 'high')? skipped - good enough
- Chinchilla optima?
  - "for compute-optimal training, the model size and
the number of training tokens should be scaled equally: for every doubling of model size the number
of training tokens should also be doubled." https://arxiv.org/pdf/2203.15556
  - Chinchilla optimal says that if you have a given compute budget, C, in FLOPs, you can train a model with M parameters on D tokens such that C ~= 6MD, so optimal training should be done with model size and token count scaling linearly with each other (and square root wrt the compute budget). "Optimal" here means "to get the lowest loss given your compute budget".
    - In other words, if we assume that training time is proportional to FLOPs, we should stick with 20 tokens per parameter to get the best loss reduction per training time. However, see the note about the 1B run in the report.
- What about things like LR?
  - Let's just use the hyperparameters from https://arxiv.org/pdf/2406.11794
    - LR=3e-3
    - Weight decay 0.033, AdamW
      - beta2=0.95 is apparently the standard for transformers: see https://github.com/mlfoundations/open_lm/
        - beta=0.9 is the default, and also appears to be standard
    - Batch size 512 (note that this agrees with our batch size=64, 1x8 configuration below)

**Report**:

According to the experiments run with the scripts in #202, 1 node with 8 h200 gpus with a batch size of 64 (effective batch size of 512) should be able to train a 100M parameter model on 2B tokens in ~6 hours using FSDP. While I could probably spend some time getting that further down, that is good enough for our purposes.

Incomplete data from the experiment (100k examples from church translation, or ~2M tokens):
```
                       104,848    1,063,776   10,397,160  101,585,792
---------------------------------------------------------------------
   1x1 GPUs  bs=1      1215.5s      1258.3s      2135.3s      3384.3s
   1x1 GPUs  bs=4       325.9s       329.5s       542.6s       891.9s
  1x1 GPUs  bs=16        88.0s        87.5s       145.0s       237.5s
  1x1 GPUs  bs=64        29.9s        29.4s        43.4s       149.8s
   1x8 GPUs  bs=1       151.2s       151.0s       251.3s       399.2s
   1x8 GPUs  bs=4        40.1s        40.4s        65.9s       107.6s
  1x8 GPUs  bs=16        11.5s        11.5s        18.6s        33.9s
  1x8 GPUs  bs=64         4.4s         4.4s         6.7s        25.9s
   2x8 GPUs  bs=1        78.4s        88.0s       255.9s      1010.5s
   2x8 GPUs  bs=4        21.8s        22.4s        65.7s       254.6s
  2x8 GPUs  bs=16         6.5s         6.8s        15.1s        66.1s
  2x8 GPUs  bs=64         2.9s         3.1s         6.5s        27.0s
   4x8 GPUs  bs=1        46.1s        61.7s       209.9s       734.6s
   4x8 GPUs  bs=4        11.9s            -            -       192.2s
  4x8 GPUs  bs=16            -         4.7s            -            -
  4x8 GPUs  bs=64         2.4s            -         5.5s        18.1s
```

<img width="1500" height="900" alt="Image" src="https://github.com/user-attachments/assets/e9333619-0b5f-47fd-b437-f92195b53370" />

Note: I don't currently have a super great guess at what it would take to train a 1B parameter model. Since I'm still running the last few timing runs, I can't currently try a 1B run. That said, between the 1M-100M range, a 10x increase in parameter count seems to roughly double the training time. If we assume training time is proportional to flops, this disagrees with chinchilla optimality! We should see a 10x increase in training time per 10x increase in parameters for the same dataset size. This means that our FLOPs are *not* proportional to training time, i.e. different configurations are giving different FLOP/s (possibly due to poor GPU utilization). For now this is not a big deal, but it might need to be addressed by #255. In the end, this might actually be a *good* thing for us, because the actual empirical multiplier is lower than the theoretical one. In theory, a full training run of a 1B parameter model would take `10*10*(training time for 100M)=600 hours=25 days`. In practice, if the ~doubling holds, it will only take ~5 days. But we need to verify that!

**Comment by JansenNye** [comment]
Very interesting that training time doesn't seem to be increasing at the theoretical rate we'd expect above 100M tokens. Looks like a good plan though

**Comment by morrisoncheney** [comment]
Splendid.
